src/python/pants/backend/rust/rules/cargo_chroot.py

```diff
@@ -0,0 +1,144 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from textwrap import dedent
+
+from pants.backend.rust.subsystems.binaries.rust_toolchain import RustToolchain
+from pants.base.build_root import BuildRoot
+from pants.build_graph.address import Address
+from pants.core.util_rules.distdir import DistDir
+from pants.engine.addresses import Addresses
+from pants.engine.cargo_intrinsics import CargoLockfile, FetchedCargoPackages
+from pants.engine.console import Console
+from pants.engine.fs import (
+    AddPrefix, Digest, DirectoryToMaterialize, FileContent, FilesContent,
+    InputFilesContent, MergeDigests, MergedResult, MergeMaybeDuplicates, RemovePrefix, PathGlobs,
+    Snapshot, SnapshotSubset, Workspace)
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import RootRule, goal_rule, rule
+from pants.engine.selectors import Get, MultiGet
+from pants.engine.target import HydrateSourcesRequest, HydratedSources, Sources, Target, Targets
+
+
+@dataclass(frozen=True)
+class SingleCargoTarget:
+    unprefixed_sources: Snapshot
+
+    @property
+    def digest(self) -> Digest:
+        return self.unprefixed_sources.digest
+
+
+@rule
+async def fetch_cargo_packages(single_target: SingleCargoTarget) -> FetchedCargoPackages:
+    lockfile_snapshot = await Get[Snapshot](SnapshotSubset(
+        digest=single_target.digest,
+        globs=PathGlobs(['Cargo.lock'])))
+    assert lockfile_snapshot.files == ('Cargo.lock',), f'snap: {lockfile_snapshot}'
+    (lockfile_content,) = tuple(await Get[FilesContent](Digest, lockfile_snapshot.digest))
+    return await Get[FetchedCargoPackages](CargoLockfile(
+        contents=lockfile_content.content.decode(),
+    ))
+
+
+@dataclass(frozen=True)
+class CargoToolchainChroot:
+    merged_digest: Digest
+
+
+@rule
+async def create_cargo_toolchain_chroot(
+    fetched_packages: FetchedCargoPackages,
+    rust_toolchain: RustToolchain,
+) -> CargoToolchainChroot:
+    crate_digests = list(fetched_packages.krate_mapping.values())
+    index_digest = fetched_packages.current_registry_index_digest
+    merged_cargo_home = await Get[Digest](MergeDigests((
+        *crate_digests,
+        index_digest,
+    )))
+    # This is intended to ensure that the CARGO_HOME from the downloaded 3rdparty deps merges with
+    # the one from the rust toolchain (which contains the cargo binary!)!
+    prefixed_cargo_home = await Get[Digest](AddPrefix(
+        digest=merged_cargo_home,
+        prefix="cargo"))
+
+    rust_toolchain_default_sentinel_file = await Get[Digest](InputFilesContent((FileContent(
+        content=f'{rust_toolchain.version}\n'.encode(),
+        path='rust-toolchain',
+    ),)))
+
+    merged = await Get[MergedResult](MergeMaybeDuplicates((
```

Comment:
It might be good for this to expose a set of paths or globs that are allowed to be duplicated... that would also serve as documentation for why the overwrites were safe.

Reply:
This is a great idea! From what I recall, the duplicated entries were 3rdparty crates with differing versions from the base installation of cargo, vs a version of that package that was in the `Cargo.lock`. I think that would be a specific subdirectory to mark. I think we probably have sufficient operations in `fs.py` to perform this algebra.